File: leetcode/0002.add-two-numbers.go
// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummpyHead := &ListNode{}
	p := dummpyHead
	var divide, sum, v1, v2 int
	for l1 != nil || l2 != nil {
		v1, v2 = 0, 0
		if l1 != nil {
			v1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			v2 = l2.Val
			l2 = l2.Next
		}
		sum = v1 + v2 + divide
		p.Next = &ListNode{Val: sum % 10}
		divide = sum / 10
		p = p.Next
	}

	if divide > 0 {
		p.Next = &ListNode{Val: divide}
	}

	return dummpyHead.Next
}

File: leetcode/0002.add-two-numbers.cpp
 struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
 };

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummpyHead;
        ListNode* p;
        p = &dummpyHead;
        int divide = 0;
        while (l1 || l2) {
            int v1 = 0;
            int v2 = 0;
            if (l1) {
                v1 = l1->val;
                l1 = l1->next;
            }
            if (l2) {
                v2 = l2->val;
                l2 = l2->next;
            }
            int sum = v1 + v2 + divide;
            divide = sum / 10;
            p->next = new ListNode{sum % 10};
            p = p->next;
        }
        if (divide > 0) {
            p->next = new ListNode{divide};
        }
        return dummpyHead.next;
    }
};
